File: packages/nanocoop-teller/src/utils/crypto.ts
import nacl from 'tweetnacl';
import { sha256 } from 'js-sha256';

export const GENESIS_HASH = '0'.repeat(64);

export function hexToBytes(hex: string): Uint8Array {
  const clean = hex.trim().replace(/^0x/, '');
  if (clean.length % 2 !== 0) {
    throw new Error(`Invalid hex length: ${clean.length}`);
  }
  const bytes = new Uint8Array(clean.length / 2);
  for (let i = 0; i < clean.length; i += 2) {
    bytes[i / 2] = parseInt(clean.substring(i, i + 2), 16);
  }
  return bytes;
}

export function bytesToHex(bytes: Uint8Array): string {
  let hex = '';
  for (let i = 0; i < bytes.length; i++) {
    hex += bytes[i].toString(16).padStart(2, '0');
  }
  return hex;
}

function formatValue(key: string, val: any): string {
  if (typeof val === 'number') {
    if (key === 'amount') {
      return Number.isInteger(val) ? val.toFixed(1) : val.toString();
    }
    return val.toString();
  }
  return JSON.stringify(val);
}

/**
 * Deterministic JSON serialization matching Python's sort_keys=True, separators=(',', ':').
 */
export function serializeForHashing(obj: any): string {
  if (obj === null || typeof obj !== 'object') {
    return JSON.stringify(obj);
  }
  if (Array.isArray(obj)) {
    return '[' + obj.map(serializeForHashing).join(',') + ']';
  }
  const keys = Object.keys(obj).sort();
  const pairs: string[] = [];
  for (const k of keys) {
    if (obj[k] !== undefined && obj[k] !== null) {
      const valStr =
        typeof obj[k] === 'object' ? serializeForHashing(obj[k]) : formatValue(k, obj[k]);
      pairs.push(JSON.stringify(k) + ':' + valStr);
    }
  }
  return '{' + pairs.join(',') + '}';
}

/**
 * Compute SHA-256(previous_hash + payload_bytes + signatures_bytes).
 */
export function generateEventHash(
  previousHash: string,
  payloadBytes: Uint8Array,
  signaturesBytes: Uint8Array
): string {
  const hasher = sha256.create();
  hasher.update(Buffer.from(previousHash, 'utf-8'));
  hasher.update(payloadBytes);
  hasher.update(signaturesBytes);
  return hasher.hex();
}

/**
 * Generate an Ed25519 keypair for Teller or Customer.
 */
export function generateKeyPair(): { privateKeyHex: string; publicKeyHex: string } {
  const keyPair = nacl.sign.keyPair();
  // tweetnacl secretKey is 64 bytes (32 bytes private seed + 32 bytes public key)
  // Python nacl uses 32 bytes private seed
  const privateSeed = keyPair.secretKey.subarray(0, 32);
  return {
    privateKeyHex: bytesToHex(privateSeed),
    publicKeyHex: bytesToHex(keyPair.publicKey),
  };
}

/**
 * Sign deterministic JSON payload with Ed25519 private seed hex.
 */
export function signPayload(privateKeyHex: string, payload: any): string {
  const seed = hexToBytes(privateKeyHex);
  const keyPair = nacl.sign.keyPair.fromSeed(seed);
  const serialized = serializeForHashing(payload);
  const msgBytes = Buffer.from(serialized, 'utf-8');
  const signature = nacl.sign.detached(msgBytes, keyPair.secretKey);
  return bytesToHex(signature);
}

/**
 * Verify Ed25519 signature over deterministic payload JSON.
 */
export function verifySignature(publicKeyHex: string, payload: any, signatureHex: string): boolean {
  try {
    const pubBytes = hexToBytes(publicKeyHex);
    const sigBytes = hexToBytes(signatureHex);
    const serialized = serializeForHashing(payload);
    const msgBytes = Buffer.from(serialized, 'utf-8');
    return nacl.sign.detached.verify(msgBytes, sigBytes, pubBytes);
  } catch {
    return false;
  }
}

File: packages/nanocoop-core/app/core/crypto.py
"""Cryptographic primitives and Ed25519 PKI utilities for NanoCoop."""

import hashlib
import json
from typing import Any
import nacl.encoding
import nacl.signing
import nacl.exceptions

GENESIS_HASH = "0" * 64


def serialize_for_hashing(payload: dict[str, Any]) -> bytes:
    """Strictly enforce sort_keys=True and separators=(',', ':') for deterministic cross-platform hashing."""
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")


def generate_event_hash(previous_hash: str, payload_bytes: bytes, signatures_bytes: bytes) -> str:
    """Compute SHA-256(previous_hash + payload + signatures)."""
    hasher = hashlib.sha256()
    hasher.update(previous_hash.encode("utf-8"))
    hasher.update(payload_bytes)
    hasher.update(signatures_bytes)
    return hasher.hexdigest()


def generate_keypair() -> tuple[str, str]:
    """Generate an Ed25519 keypair. Returns (private_key_hex, public_key_hex)."""
    signing_key = nacl.signing.SigningKey.generate()
    private_hex = signing_key.encode(encoder=nacl.encoding.HexEncoder).decode("utf-8")
    public_hex = signing_key.verify_key.encode(encoder=nacl.encoding.HexEncoder).decode("utf-8")
    return private_hex, public_hex


def sign_payload(private_key_hex: str, payload: dict[str, Any]) -> str:
    """Sign deterministic JSON payload with Ed25519 private key. Returns signature hex string."""
    signing_key = nacl.signing.SigningKey(
        private_key_hex.encode("utf-8"), encoder=nacl.encoding.HexEncoder
    )
    payload_bytes = serialize_for_hashing(payload)
    signed = signing_key.sign(payload_bytes)
    return signed.signature.hex()


def verify_signature(public_key_hex: str, payload: dict[str, Any], signature_hex: str) -> bool:
    """Verify Ed25519 signature over deterministic payload JSON."""
    try:
        verify_key = nacl.signing.VerifyKey(
            public_key_hex.encode("utf-8"), encoder=nacl.encoding.HexEncoder
        )
        payload_bytes = serialize_for_hashing(payload)
        signature_bytes = bytes.fromhex(signature_hex)
        verify_key.verify(payload_bytes, signature_bytes)
        return True
    except (nacl.exceptions.BadSignatureError, ValueError, Exception):
        return False


def hash_leaf(data: str) -> str:
    """Compute SHA-256 leaf hash for Merkle tree."""
    return hashlib.sha256(data.encode("utf-8")).hexdigest()


def hash_pair(left: str, right: str) -> str:
    """Compute SHA-256 of combined parent node."""
    hasher = hashlib.sha256()
    hasher.update(left.encode("utf-8"))
    hasher.update(right.encode("utf-8"))
    return hasher.hexdigest()


def compute_merkle_root(leaf_hashes: list[str]) -> str:
    """Compute the Merkle Tree root hash from an ordered list of leaf event hashes.

    If empty, returns GENESIS_HASH.
    If odd number of leaves, the last leaf is duplicated to form a balanced tree.
    """
    if not leaf_hashes:
        return GENESIS_HASH

    current_layer = [hash_leaf(h) for h in leaf_hashes]

    while len(current_layer) > 1:
        next_layer: list[str] = []
        for i in range(0, len(current_layer), 2):
            left = current_layer[i]
            right = current_layer[i + 1] if i + 1 < len(current_layer) else left
            next_layer.append(hash_pair(left, right))
        current_layer = next_layer

    return current_layer[0]
